azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -107,12 +99,441 @@ def mock_execute_op(self, op):
+        stage.pipeline_root.connected = False
+
+        stage.run_op(op)
+        connect_op = stage.next.run_op.call_args[0][0]
+        operation_flow.complete_op(stage=stage.next, op=connect_op)
+        op.error = fake_exception
+        operation_flow.complete_op(stage=stage.next, op=op)
+
+        assert_callback_failed(op=op, error=fake_exception)
+
+
+pipeline_stage_test.add_base_pipeline_stage_tests(
+    cls=pipeline_stages_base.SerializeConnectOpsStage,
+    module=this_module,
+    all_ops=all_common_ops,
+    handled_ops=[
+        pipeline_ops_base.ConnectOperation,
+        pipeline_ops_base.DisconnectOperation,
+        pipeline_ops_base.ReconnectOperation,
+    ],
+    all_events=all_common_events,
+    handled_events=[],
+    extra_initializer_defaults={"blocked": False, "queue": queue.Queue},
+)
+
+connect_ops = [
+    {"op_class": pipeline_ops_base.ConnectOperation, "connected_flag_required_to_run": False},
+    {"op_class": pipeline_ops_base.DisconnectOperation, "connected_flag_required_to_run": True},
+    {"op_class": pipeline_ops_base.ReconnectOperation, "connected_flag_required_to_run": True},
+]
+
+
+class FakeOperation(pipeline_ops_base.PipelineOperation):
+    pass
+
+
+@pytest.mark.describe("SerializeConnectOpsStage - .run_op() -- called with connect operation")
+class TestSerializeConnectOpStageRunOp(object):
+    @pytest.fixture
+    def stage(self, mocker):
+        stage = make_mock_stage(
+            mocker=mocker, stage_to_make=pipeline_stages_base.SerializeConnectOpsStage
+        )
+        stage.next.run_op = mocker.MagicMock()
+        return stage
+
+    @pytest.fixture
+    def connect_op(self, mocker, params):
+        return params["op_class"](callback=mocker.MagicMock())
+
+    @pytest.fixture
+    def fake_op(self, mocker):
+        return FakeOperation(callback=mocker.MagicMock())
+
+    @pytest.fixture
+    def fake_ops(self, mocker):
+        return [
+            FakeOperation(callback=mocker.MagicMock()),
+            FakeOperation(callback=mocker.MagicMock()),
+            FakeOperation(callback=mocker.MagicMock()),
+        ]
+
+    @pytest.mark.it(
+        "Immediately completes a ConnectOperation if the transport is already connected"
+    )
+    def test_connect_while_connected(self, stage, mocker):
+        op = pipeline_ops_base.ConnectOperation(callback=mocker.MagicMock())
+        stage.pipeline_root.connected = True
+        stage.run_op(op)
+        assert_callback_succeeded(op=op)
+
+    @pytest.mark.it(
+        "Immediately completes a DisconnectOperation if the transport is already disconnected"
+    )
+    def test_disconnect_while_disconnected(self, stage, mocker):
+        op = pipeline_ops_base.DisconnectOperation(callback=mocker.MagicMock())
+        stage.pipeline_root.connected = False
+        stage.run_op(op)
+        assert_callback_succeeded(op=op)
+
+    @pytest.mark.it(
+        "Immediately passes the operation down if an operation is not alrady being serialized"
```

Comment:
>serialized [](start = 83, length = 10)

Probably just my lack of knowledge, but I'm unclear exactly what `serialized` means in this context. I get what this is referring to - an already running connection operation. blocking all other operations, I just don't see how that means the operation is "being serialized".

Also probably worth specifying "another connection operation is not already being serialized", since only connection operations can block.

Reply:
yes, this is confusing.  I changed all the mark.it text to talk about operations that are blocking the stage.  I think this is much clearer.

---
In reply to: [312674162](https://github.com/Azure/azure-iot-sdk-python-preview/pull/160#discussion_r312674162) [](ancestors = 312674162)